inference-engine/src/vpu/common/src/ngraph/transformations/merge_gather_gather_elements.cpp

```diff
@@ -0,0 +1,156 @@
+        };
+
+        if (!std::all_of(transposeConsumers.cbegin(), transposeConsumers.cend(), isShapeOfOrGatherElements)) {
+            continue;
+        }
+
+        for (const auto& transposeConsumer : transposeConsumers) {
+            if (transposeConsumer.get_node()->get_type_info() == ngraph::opset6::ShapeOf::type_info) {
+                shapeOfs.push_back(transposeConsumer.get_node());
+            } else if (transposeConsumer.get_node()->get_type_info() == ngraph::opset6::GatherElements::type_info) {
+                gatherElements.push_back(transposeConsumer.get_node());
+            }
+        }
+
+        const auto& matchedTransposePerm = patternMap.at(transposePerm);
+        const auto& matchedSqueezeAxes = patternMap.at(squeezeAxes);
+        const auto& matchedSqueeze = patternMap.at(squeeze);
+        const auto& matchedGatherData = patternMap.at(gatherData);
+        const auto& matchedGatherIndices = patternMap.at(gatherIndices);
+        const auto& matchedGather = patternMap.at(gather);
+
+        for (const auto& gatherElement : gatherElements) {
+            const auto transposeIndices = std::make_shared<ngraph::opset6::Transpose>(gatherElement->input_value(1), matchedTransposePerm);
+            const auto unsqueezeIndices = std::make_shared<ngraph::opset6::Unsqueeze>(
+                transposeIndices,
+                matchedSqueezeAxes);
+            const auto expGatherElements = std::make_shared<ngraph::vpu::op::ExpGatherElements>(
+                matchedGatherData,
+                unsqueezeIndices,
+                matchedGatherIndices,
+                ngraph::as_type<ngraph::opset6::GatherElements>(gatherElement)->get_axis(),
+                ngraph::as_type<ngraph::opset6::Gather>(matchedGather.get_node())->get_axis());
+            const auto squeezeData = matchedSqueeze.get_node()->clone_with_new_inputs({expGatherElements, matchedSqueezeAxes});
+            const auto transposeData = matchedTranspose.get_node()->clone_with_new_inputs({squeezeData, matchedTransposePerm});
+            transposeData->set_friendly_name(gatherElement->get_friendly_name());
+            gatherElement->get_default_output().replace(transposeData);
+            wasGraphChanged = true;
+        }
+
+        for (const auto& shapeOf : shapeOfs) {
+            const auto gatherDataShape = std::make_shared<ngraph::opset6::ShapeOf>(
+                matchedGatherData,
+                ngraph::as_type<ngraph::opset6::ShapeOf>(shapeOf)->get_output_type());
+            const auto gatherIndicesShape = std::make_shared<ngraph::opset6::ShapeOf>(
+                matchedGatherIndices,
+                ngraph::as_type<ngraph::opset6::ShapeOf>(shapeOf)->get_output_type());
+            const auto axis = ngraph::as_type<ngraph::opset6::Gather>(matchedGather.get_node())->get_axis();
+
+            const auto gatherIndicesRank = matchedGatherIndices.get_partial_shape().rank().get_length();
```

Comment:
You need to check that rank is static.

Reply:
Added `pattern::has_static_rank()` to the Matcher.